docs/design/document-type.md

```diff
@@ -14,346 +14,164 @@ such as `struct`, `union`, etc. which are handled by the [Kotlin Smithy SDK](kot
-     */
-    fun getMap(index: Int) = content[index] as? SmithyMap
-
-    override fun toString(): String = content.joinToString( separator = ",", prefix = "[", postfix = "]" )
-}
-
-/**
- * Class representing document `map` type
- *
- * Map consists of name-value pairs, where the value is an arbitrary Document. This is much like a JSON object.
- */
-data class SmithyMap(val content: Map<String, Document>): Document(), Map<String, Document> by content {
-    /**
-     * Returns [SmithyNumber] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getNumber(key: String): SmithyNumber? = getValue(key) as? SmithyNumber
-
-    /**
-     * Returns [SmithyBool] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getBoolean(key: String): SmithyBool? = getValue(key) as? SmithyBool
-
-    /**
-     * Returns [SmithyString] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getString(key: String): SmithyString? = getValue(key) as? SmithyString
-
-    /**
-     * Returns [SmithyArray] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getArray(key: String): SmithyArray? = getValue(key) as? SmithyArray
-
-    /**
-     * Returns [SmithyMap] associated with given [key] or `null` if element is not present or has a different type
-     */
-    fun getMap(key: String): SmithyMap? = getValue(key) as? SmithyMap
-
-    override fun toString(): String {
-        return content.entries.joinToString(
-            separator = ",",
-            prefix = "{",
-            postfix = "}",
-            transform = {(k, v) -> """"$k":$v"""}
-        )
-    }
-}
-
-fun Boolean.toDocument() = SmithyBool(this)
-fun Number.toDocument() = SmithyNumber(this)
-fun String.toDocument() = SmithyString(this)
-
-/**
- * DSL builder for a [SmithyArray]
- */
-class DocumentArrayBuilder internal constructor() {
-    internal val content: MutableList<Document> = mutableListOf()
-
-    /**
-     * Adds [this] value to the current [SmithyArray] as [SmithyString]
-     */
-    operator fun String.unaryPlus() {
-        content.add(SmithyString(this))
+    companion object {
+        /**
+         * Overloads to create a document from possible values.
+         */
+        operator fun invoke(init: Builder.() -> Unit): Document
+        operator fun invoke(value: kotlin.Number): Document
+        operator fun invoke(value: kotlin.String): Document
+        operator fun invoke(value: kotlin.Boolean): Document
+        operator fun invoke(value: Transformable): Document
+
+        /**
+         * Unwraps a Document from a serialized value.
+         */
+        fun fromString(value: kotlin.String): Document
+        
+        /**
+         * Creates a Document list from the provided values.
+         */
+        fun listOf(vararg values: Any?): Document
```

Comment:
**Comment**: It's a shame we can't provide a narrower parameter list than this. If Kotlin ever supports union types this might be good place to use them.

Reply:
Theoretically we could, though not quite in the same saw - I was looking at kotlinx serialization and this is a sample of their DSL builder for json objects:

```kotlin
val json = buildJsonObject {
    put("booleanKey", true)
    putJsonArray("arrayKey") {
        for (i in 1..10) add(i)
    }
    putJsonObject("objectKey") {
        put("stringKey", "stringValue")
    }
}
```
If we do it this way, all of the builder members could be defined as typed overloads (add number, add string, etc) and I'd still consider it at least somewhat "idiomatic" considering we're following how kotlinx does a thing.